leo/package/src/program.rs

```diff
@@ -0,0 +1,130 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::*;
+
+use leo_errors::{PackageError, Result, UtilError};
+use leo_span::Symbol;
+
+use snarkvm::prelude::{Program as SvmProgram, TestnetV0};
+
+use indexmap::IndexSet;
+use std::path::Path;
+
+/// Information about an Aleo program.
+#[derive(Clone, Debug)]
+pub struct Program {
+    // The name of the program (no ".aleo" suffix).
+    pub name: Symbol,
+    pub data: ProgramData,
+    pub dependencies: IndexSet<Dependency>,
+}
+
+impl Program {
+    /// Given the location `path` of a local Leo package, read the filesystem
+    /// to obtain a `Program`.
+    pub fn from_path<P: AsRef<Path>>(name: Symbol, path: P) -> Result<Self> {
+        Self::from_path_impl(name, path.as_ref())
+    }
+
+    fn from_path_impl(name: Symbol, path: &Path) -> Result<Self> {
+        let manifest = Manifest::read_from_file(path.join(MANIFEST_FILENAME))?;
+        let manifest_symbol = crate::symbol(&manifest.program)?;
+        if name != manifest_symbol {
+            return Err(PackageError::conflicting_manifest(
+                format_args!("{name}.aleo"),
+                format_args!("{manifest_symbol}.aleo"),
+            )
+            .into());
+        }
+        let source_directory = path.join(SOURCE_DIRECTORY);
+        let count = source_directory
+            .read_dir()
+            .map_err(|e| {
+                UtilError::util_file_io_error(
+                    format_args!("Failed to read directory {}", source_directory.display()),
+                    e,
+                )
+            })?
+            .count();
+
+        let source_path = source_directory.join(MAIN_FILENAME);
+
+        if !source_path.exists() || count != 1 {
+            return Err(PackageError::source_directory_can_contain_only_one_file(source_directory.display()).into());
+        }
+
+        Ok(Program {
+            name,
+            data: ProgramData::SourcePath(source_path),
+            dependencies: manifest.dependencies.unwrap_or_default().into_iter().collect(),
+        })
+    }
+
+    /// Given an Aleo program on a network, fetch it to build a `Program`.
+    pub fn fetch<P: AsRef<Path>>(name: Symbol, home_path: P, network: NetworkName, endpoint: &str) -> Result<Self> {
+        Self::fetch_impl(name, home_path.as_ref(), network, endpoint)
+    }
+
+    fn fetch_impl(name: Symbol, home_path: &Path, network: NetworkName, endpoint: &str) -> Result<Self> {
```

Comment:
Should this be merged with the call to fetch_impl?

Reply:
But `fetch_impl` isn't called anywhere else, so couldn't the body be merged with something like `home_path = home_path.as_ref()` at the top?

Reply:
Ah good point! I've never had to seriously consider binary size in my Rust development but I like the reasoning.